include/net/net_ip.h

```diff
@@ -162,12 +162,8 @@ struct sockaddr_storage {
 struct net_addr {
 	sa_family_t family;
 	union {
-#if defined(CONFIG_NET_IPV6)
```

Comment:
I think this is bad idea because it uses more memory now. Even thou using #ifdefs looks ugly, there is a place for them here.

Reply:
+1, Except there's nothing "ugly" in #ifdef's, they're beautiful. There's a way to make them even more beautiful though - indent them, just like normal code is being indented. (Well, some conventions should be set, e.g. here "#if" could start either at "union" indent or "struct in6_addr in6_addr" indent).

Reply:
#ifdefs looking ugly is not the worst thing. Dead code is. With ifdefs you can have a lot of code which is not touched by the compiler until someone, someday enables the feature and finds out that it does not compile.

Personally, i tried to avoid #ifdefs as much as possible and there are ways to keep memory usage low.

In that case it's all about the size of struct in6_addr so it could be reduced to a pointer:
```
union {
#if defined(CONFIG_NET_IPV6)
		u8_t u6_addr8[16];
		u16_t u6_addr16[8]; /* In big endian */
		u32_t u6_addr32[4]; /* In big endian */
#else 
		u8_t *u6_addr8;
		u16_t *u6_addr16; /* In big endian */
		u32_t *u6_addr32; /* In big endian */
#endif 
	} in6_u;
```

Reply:
Let's try with that.

Reply:
> Except there's nothing "ugly" in #ifdef's, they're beautiful.

YMMV, but #ifdefs hide away code paths (as nordic-krch pointed out) and are therefore an unnecessary evil to wrestle with. And the issue two levels up explicitly asked for #ifdef removal.

Reply:
@nordic-krch 

> Dead code is. With ifdefs you can have a lot of code which is not touched by the compiler until someone, someday enables the feature and finds out that it does not compile.

That's very obvious problem with very obvious solution - have a good CI with good coverage. Zephyr exactly has that. Otherwise, saying that is about the same as saying "if's are bad - suppose some branch is not executed, until someone, someday. And then [fighters flip upside down on crossing equator](https://forums.theregister.co.uk/forum/1/2007/02/28/f22s_working_again/)". Solution is not killing *all* if's, but again, have good testing coverage.

> Personally, i tried to avoid #ifdefs as much as possible and there are ways to keep memory usage low.

Of course, everyone uses #ifdef's only when required, and doesn't use when not required.



Now let's look at the risks of using code like `if (1)`/`if (0)` (to which IS_ENABLED translates):

* It's effectively dead code, some compilers love to throw warnings at such, which can be turned to errors easily.
* Some overzealous static analysis tools can do the same. Coverity which we use seems to be smart not to (or was ~~patched~~ configured not to), but who knows what will be used next.
* Finally, there're creative coding style standards. One of the most creative known is MISRA C. With all the mad things it required, it would be a wonder if it allowed dead if's.


So, with no-zero probability, all this fashion-motion to go from #ifdef to if(IS_ENABLED()) can go into reverse direction.